numpy/_core/src/multiarray/item_selection.c

```diff
@@ -3085,43 +3103,68 @@ PyArray_Sort(PyArrayObject *op, int axis, NPY_SORTKIND flags)
     // Zero the NPY_HEAPSORT bit, maps NPY_HEAPSORT to NPY_QUICKSORT
     flags &= ~_NPY_SORT_HEAPSORT;
 
-    sort_table = PyDataType_GetArrFuncs(PyArray_DESCR(op))->sort;
-    switch (flags) {
-        case NPY_SORT_DEFAULT:
-            sort = sort_table[NPY_QUICKSORT];
-            break;
-        case NPY_SORT_STABLE:
-            sort = sort_table[NPY_STABLESORT];
-            break;
-        default:
-            break;
-    }
-
-    // Look for appropriate generic function if no type specific version
-    if (sort == NULL) {
-        if (!PyDataType_GetArrFuncs(PyArray_DESCR(op))->compare) {
-            PyErr_SetString(PyExc_TypeError,
-                            "type does not have compare function");
+    // Look for type specific functions
+    sort_method = NPY_DT_SLOTS(NPY_DTYPE(PyArray_DESCR(op)))->sort_meth;
+    if (sort_method != NULL) {
+        PyArrayMethod_SortParameters sort_params = {
+            .flags = flags,
+        };
+        PyArray_Descr *descr = PyArray_DESCR(op);
+        _context.descriptors = (PyArray_Descr * const[]){descr, descr};
```

Comment:
Should this be a single one? Isn't sorting in-place.

So, ufuncs have `resolve_descriptors()` and if we use that, then I think it makes sense to do it here.
That replaces the later call to make the dtype native byte order or canonical.

Reply:
Aren't array method loops defined to have `data[nin+nout]`? I don't think there is any place they are defined to be in-place, and it would seem weird to interpret `nin=1`, `nout=0` as in-place, especially as that wouldn't scale to multiple in-place operations. It seems safer to just recognize in-place inside the loop with `data[0] == data[1]` (same strides are then guaranteed [and of course certainly here]). Of course, that has the additional advantage that it would be the signature a potential future gufunc would require...

Reply:
Fair, just yet another special case if that is always the case for sort, thought one nout might be fine (or even in/out if ever a gufunc). Because if a gufunc, then you'll have to support non-inplace as well.

Reply:
Since over in `erfa`, I have a gufunc (yes, a crazy one, just because it was easier to wrap every C bit in gufuncs that special-case something) that has `nin=0, nout=1`, I don't think we should presume how to signal in-place! 

Anyway, to me perhaps too many functions look like `gufunc`, but I do hope to have a try at wrangling sorting and partitioning etc into some and it would be nice if we can ensure that if that ever were to happen, anybody who had written sort loops for the mechanism here would not have to change a thing.